shell/common/rasterizer.cc

```diff
@@ -120,50 +119,32 @@ void Rasterizer::Draw(fml::RefPtr<Pipeline<flutter::LayerTree>> pipeline) {
   }
   FML_DCHECK(task_runners_.GetRasterTaskRunner()->RunsTasksOnCurrentThread());
 
-  RasterStatus raster_status = RasterStatus::kFailed;
-  Pipeline<flutter::LayerTree>::Consumer consumer =
-      [&](std::unique_ptr<LayerTree> layer_tree) {
-        raster_status = DoDraw(std::move(layer_tree));
-      };
-
-  PipelineConsumeResult consume_result = pipeline->Consume(consumer);
-  // if the raster status is to resubmit the frame, we push the frame to the
-  // front of the queue and also change the consume status to more available.
-  if (raster_status == RasterStatus::kResubmit) {
-    auto front_continuation = pipeline->ProduceIfEmpty();
-    bool result =
-        front_continuation.Complete(std::move(resubmitted_layer_tree_));
-    if (result) {
-      consume_result = PipelineConsumeResult::MoreAvailable;
-    }
-  } else if (raster_status == RasterStatus::kEnqueuePipeline) {
-    consume_result = PipelineConsumeResult::MoreAvailable;
-  }
+  std::unique_ptr<LayerTree> layer_tree = layer_tree_holder->Get();
+  RasterStatus raster_status =
+      layer_tree ? DoDraw(std::move(layer_tree)) : RasterStatus::kFailed;
 
   // Merging the thread as we know the next `Draw` should be run on the platform
   // thread.
   if (raster_status == RasterStatus::kResubmit) {
+    layer_tree_holder->ReplaceIfNewer(std::move(resubmitted_layer_tree_));
     auto* external_view_embedder = surface_->GetExternalViewEmbedder();
-    // We know only the `external_view_embedder` can
-    // causes|RasterStatus::kResubmit|. Check to make sure.
-    FML_DCHECK(external_view_embedder != nullptr);
+    FML_DCHECK(external_view_embedder != nullptr)
+        << "kResubmit is an invalid raster status without external view "
+           "embedder.";
     external_view_embedder->EndFrame(raster_thread_merger_);
   }
 
-  // Consume as many pipeline items as possible. But yield the event loop
+  // Consume as many layer trees as possible. But yield the event loop
   // between successive tries.
-  switch (consume_result) {
-    case PipelineConsumeResult::MoreAvailable: {
-      task_runners_.GetRasterTaskRunner()->PostTask(
-          [weak_this = weak_factory_.GetTaskRunnerAffineWeakPtr(), pipeline]() {
-            if (weak_this) {
-              weak_this->Draw(pipeline);
-            }
-          });
-      break;
-    }
-    default:
-      break;
+  // Note: This behaviour is left as-is to be inline with the pipeline
+  // semantics. TODO(kaushikiska): explore removing this block.
+  if (!layer_tree_holder->IsEmpty()) {
```

Comment:
Whats making this method confusing to me is the re-submission logic. IMO, we should get rid of the `resubmitted_layer_tree_` and have the caller to DoDraw (this frame) insert the layer tree that because of resubmit insert that tree back into the holder. We can then make DoDraw const so we its easier to read.

I believe we can get rid of this block if we say that whoever pushes a layertree to holder also has to perform a draw explicitly.

Reply:
There is an entire comment section below about how thread merging interacts with pipeline pressure that is now irrelevant and confusing(GitHub wont allow me to comment on that line in this patch [but the link is here](https://github.com/flutter/engine/blob/8b4987382eddeaf4dbcaf927ab6b4b24e2e3b3a5/shell/common/rasterizer.cc#L315)). I think we need to do a review of the documentation about to make sure outdated concepts are removed. Specifically, we need to make sure we have consistent documentation that discusses the following edge cases:

1: What happens when there is frame re-resubmission because of thread merging and unmerging?
2: What happens when new layer tree are not being produced in case of external texture composition?

Fortunately, this new world is simpler as there is no concept of pipeline backpressure.

Reply:
I 100% agree with this. One of the motivations for me to remove the pipeline mechanism is to facilitate easier thread merging and resubmission. I'd like to keep the thread merging behavior exactly the same for this commit and will simplify the resubmission logic along with clearing up some of the `RasterStatus` enums that are no longer valid.